### ISSUE TITLE
Rule proposal: `no-array-fill-with-reference-type`

### ISSUE THREAD

**Issue by legend80s** [sindresorhus/eslint-plugin-unicorn#2657]
## **Rule: `no-array-fill-with-reference-type`**  
**Description**:  
Disallows using `Array.fill()` with **reference types** (objects, arrays, functions, Maps, Sets, RegExp literals, etc.) to prevent unintended shared references across array elements. Encourages `Array.from()` or explicit iteration for creating independent instances.  

**Reason**:  
`Array(len).fill(value)` fills all elements with the **same reference** if `value` is non-primitive (e.g., `fill([])`), leading to bugs when one element’s mutations affect others.  

**Key Features**:  
- Catches **all reference types**: Objects, Arrays, Functions, `new` expressions, RegExp literals, and variables referencing them.  
- **Clear error messages**: Identifies the reference type (e.g., `Object`, `RegExp`, `variable (name)`).  
- **Recommends fixes**: Suggests `Array.from()` or similar patterns.  

**Note**: Primitive types (`number`, `string`, `boolean`, `null`, `undefined`, `symbol`, `bigint`) are always allowed.

Here's a polished version of your text with improved clarity and flow:

---

## Why Existing Rules Don't Address This Issue

The [`no-new-array`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-new-array.md) rule prohibits using `new Array()` or `Array()` (maybe) constructor calls, but it doesn't specifically address the reference-sharing problem with `.fill()`. Since developers may disable this rule when they legitimately need array constructors, we need a dedicated rule to:

1. **Target the specific hazard** of reference sharing when using `.fill()`
2. **Provide focused guidance** on safe alternatives like `Array.from()`

This specialized rule complements rather than replaces `no-new-array`, offering more granular protection against this common pitfall.

---
## Need fix?

No fix needed because this rule is automatically fixable by `no-new-array`. or the `fix` logic can use the same as `no-new-array`.

### Examples

```js
// ❌
new Array(3).fill({});       // ✗ Object  
new Array(3).fill(new Map());       // ✗ Map
new Array(3).fill(new Set());       // ✗ Set
new Array(3).fill(/pattern/); // ✗ RegExp 
new Array(3).fill(new String('fff'));       // ✗ new String
new Array(3).fill(new Foo('fff'));       // ✗ new Class
new Array(3).fill(Class);       // ✗ Class or function

const map = new Map();  
new Array(3).fill(map);      // ✗ Variable (map)  

// ✅
new Array(3).fill(0);        // ✓ number (primitive)  
new Array(3).fill(10n);        // ✓ bigint (primitive)  
new Array(3).fill(null);     // ✓ null (primitive)  
new Array(3).fill(undefined);     // ✓ undefined(primitive)  
new Array(3).fill('foo');        // ✓ string (primitive)  
new Array(3).fill(false);     // ✓ boolean (primitive)  
new Array(3).fill(Symbol('foo'));        // ✓ Symbol(primitive)  

Array.from({ length: 3 }, () => ({})); // ✓ Safe alternative
```

### Proposed rule name

no-array-fill-with-reference-type

**Comment by legend80s** [comment]
Can I make a PR for this?
